### PR TITLE
Adds Cohort Characterization stratification

### PR DESCRIPTION
**Pull request by wivern**
add stratification for #649

### CODE REVIEW

File: src/main/java/org/ohdsi/webapi/cohortcharacterization/GenerateCohortCharacterizationTasklet.java [hunk 1]
```diff
@@ -268,50 +287,118 @@ private boolean ccHasPresetDistributionAnalyses() {
...
+              } else {
+                throw new IllegalArgumentException(String.format("Feature class %s is not supported", feature.getClass()));
+              }
+            } else {
+              throw new IllegalArgumentException(String.format("Stat type %s is not supported", analysis.getStatType()));
+            }
+            return groupQuery;
+        }
+
+        private List<PreparedStatementCreator> getQueriesForStratifiedCriteriaAnalyses(Integer cohortDefinitionId) {
```
Comment (pavgra) [inline src/main/java/org/ohdsi/webapi/cohortcharacterization/GenerateCohortCharacterizationTasklet.java]: method `getFeatureAnalysisQueries` used in the function is creating not only Criteria analyses queries, so the naming is confusing

File: src/main/java/org/ohdsi/webapi/cohortcharacterization/GenerateCohortCharacterizationTasklet.java [hunk 2]
```diff
@@ -348,30 +438,52 @@ private PreparedStatementRendererCreator prepareStatement(String query, String s
...
             List<PreparedStatementCreator> queriesToRun = new LinkedList<>();
 
-            List<PreparedStatementCreator> createQueries = Arrays.stream(SqlSplit.splitSql(jsonObject.getString("sqlConstruction")))
+            queriesToRun.addAll(getCreateQueries(jsonObject));
+
+            queriesToRun.addAll(getFeatureAnalysisQueries(jsonObject, cohortDefinitionId, null));
+
+            queriesToRun.addAll(getCleanupQueries(jsonObject));
+
+            if (!cohortCharacterization.getStratas().isEmpty()) {
```
Comment (pavgra) [inline src/main/java/org/ohdsi/webapi/cohortcharacterization/GenerateCohortCharacterizationTasklet.java]: So, if CC is stratified, you will have all analyses running twice? First, `getFeatureAnalysisQueries` is called at line 443, then - inside of `getQueriesForStratifiedCriteriaAnalyses`

File: src/main/java/org/ohdsi/webapi/cohortcharacterization/GenerateCohortCharacterizationTasklet.java (hunk 1 above)
Comment (wivern) [reply src/main/java/org/ohdsi/webapi/cohortcharacterization/GenerateCohortCharacterizationTasklet.java]: probably getFeatureAnalys**e**sQueries is more suitable, and on my taste FeatureAnalyses is not only about criteria analyses.
Could you suggest a better name?

File: src/main/java/org/ohdsi/webapi/cohortcharacterization/GenerateCohortCharacterizationTasklet.java (hunk 2 above)
Comment (wivern) [reply src/main/java/org/ohdsi/webapi/cohortcharacterization/GenerateCohortCharacterizationTasklet.java]: not twice, analyses running on the different target cohort tables

File: src/main/resources/resources/cohortcharacterizations/sql/strataWithCriteria.sql [hunk 3]
```diff
@@ -0,0 +1,11 @@
+WITH qualified_events AS (
+    SELECT ROW_NUMBER() OVER (partition by E.subject_id order by E.cohort_start_date) AS event_id, E.subject_id AS person_id, E.cohort_start_date AS start_date, E.cohort_end_date AS end_date, OP.observation_period_start_date AS op_start_date, OP.observation_period_end_date AS op_end_date
+    FROM @temp_database_schema.@targetTable E
+      JOIN @cdm_database_schema.observation_period OP ON E.subject_id = OP.person_id AND E.cohort_start_date >= OP.observation_period_start_date AND E.cohort_start_date <= OP.observation_period_end_date
+    WHERE cohort_definition_id = @cohortId
+)
+INSERT INTO @strataCohortTable(cohort_definition_id, strata_id, subject_id, cohort_start_date, cohort_end_date)
```
Comment (pavgra) [inline src/main/resources/resources/cohortcharacterizations/sql/strataWithCriteria.sql]: https://docs.microsoft.com/en-us/sql/t-sql/queries/with-common-table-expression-transact-sql?view=sql-server-2017

> Features and Limitations of Common Table Expressions in SQL Data Warehouse and Parallel Data Warehouse
> A CTE must be followed by a single SELECT statement. INSERT, UPDATE, DELETE, and MERGE statements are not supported.

**Comment by JamesSWiggins** [comment]
I now get the below errors when trying to build the branch "v2.6.0"

[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /tmp/WebAPI/src/main/java/org/ohdsi/webapi/cohortcharacterization/domain/CohortCharacterizationEntity.java:[16,7] error: CohortCharacterizationEntity is not abstract and does not override abstract method getStratas() in CohortCharacterization
[ERROR] /tmp/WebAPI/src/main/java/org/ohdsi/webapi/cohortcharacterization/dto/CcExportDTO.java:[7,7] error: CcExportDTO is not abstract and does not override abstract method getStratas() in CohortCharacterization
[INFO] 2 errors
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE

**Comment by anthonysena** [comment]
@JamesSWiggins I think this may be due to some updates on a dependent library here: https://github.com/OHDSI/StandardizedAnalysisAPI/commit/680c08214046a5a7edc66b0a044d6a02064a8775. 

@pavgra: Should the StandardizedAnalysisAPI get a revision (1.0.1) or perhaps we can coordinate an update to the Snapshot on the OHDSI nexus repo?

**Comment by pavgra** [comment]
@anthonysena , ouch, yes, we had to increase the version to `1.0.0-SNAPSHOT`

**Comment by anthonysena** [comment]
@pavgra - can you tag the releases on StandardizedAnalysisAPI? I noticed we did not formally do a v1.0.0 of StandardizedAnalysisAPI when we released WebAPI v2.6.0 and we should make sure that there is some linkage there. We can then work with Lee to get things properly versioned for Nexus.

**Comment by JamesSWiggins** [comment]
Is this something I need to change in my compile environment?  Still getting the same error on the v2.6.0 branch.

**Comment by anthonysena** [comment]
@JamesSWiggins - sorry for the delay. We've gotten the versioning issues all sorted out if you'd like to give it a go. Just let us know if you encounter any issues. Thanks!

**Comment by JamesSWiggins** [comment]
@anthonysena @pavgra @wivern Thanks for fixing this!!  Builds well for me now and also is working with my [Atlas/WebAPI upgrade automation for AWS](https://github.com/JamesSWiggins/ohdsi-cfn#upgrading-atlaswebapi).

**Pull request merged**